### PR TITLE
add font rendering options

### PR DESCRIPTION
**Pull request by ali65**
On low resolution LCD, with the right font type and the hinting method, there is a better looking result if antialiasing is disabled. With retina display, this is not an issue, but not all kivy users have retina.

Kivy is hardwired to enable antialiasing because it always calls TTF_RenderUTF8_Blended.
With this change Kivy could optionally call the TTF_RenderUTF8_Solid (which does not use antialiasing).

Beside that the change allows setting the hinting level.

Additional info:
http://www.sdltutorials.com/sdl-ttf

Further thinking: Maybe kivy should automatically detect if display is low res and font type is such, and switch between antialising enabled/disabled automatically? For simplicity it is better just leave control to the programmer.

### CODE REVIEW

**Comment by dessant** [comment]
Nice features. :+1: 
I have a minor reservation about the hinting api though, could you make it an OptionProperty?
https://kivy.org/docs/api-kivy.properties.html#kivy.properties.OptionProperty

The options could be: `None`, `'mono'`, `'light'`, `'normal'`.

The new api endpoints will also need a bit of docs, please follow this as a template:
https://github.com/udiboy1209/kivy/blob/styling/kivy/uix/label.py#L641.

Also add a note to them about sdl2, like this:
https://github.com/kivy/kivy/blob/4e429deb8864e3450cb71cd670cd00d1516d2799/kivy/core/window/__init__.py#L837.

**Comment by ali65** [comment]
Shortly I will correct the issues, you have noticed.
I am glad, not only I think; this feature would be nice to add.

**Comment by dessant** [comment]
Thanks a lot for your work. Yes, absolutely, we intend to merge it. :wink:

**Comment by matham** [comment]
Since you're still making fixes, it may be nice if you explain in the docs what kerning, hinting and blending do. Since I for one don't know what hinting or blending does. It should just be a short one or two sentences.

File: kivy/uix/label.py [hunk 1]
```diff
@@ -812,3 +813,38 @@ def print_it(instance, value):
...
     '''
+
+    font_hinting = OptionProperty(
+        'none', options=['normal', 'light', 'mono', 'none'])
+    '''What hinting option to use for font rendering.
+    Can be `'normal'`, `'light'`, `'mono'`, `'none'`
+
+    .. note::
+        This feature requires a SDL2 window provider and is currently only
+        supported on desktop platforms.
```
Comment (akshayaurora) [inline kivy/uix/label.py]: sadly is now being supported everywhere even on mobiles.

Comment (akshayaurora) [inline kivy/uix/label.py]: ..versionadded::1.9.2

File: kivy/uix/label.py [hunk 2]
```diff
@@ -812,3 +813,38 @@ def print_it(instance, value):
...
+    defaults to `'none'`.
+    '''
+
+    font_kerning = BooleanProperty(False)
+    '''Whether kerning is enable for font rendering
+
+    .. note::
+        This feature requires a SDL2 window provider and is currently only
+        supported on desktop platforms.
+
```
Comment (akshayaurora) [inline kivy/uix/label.py]: Same as above, remove the warning about desktops and include ..versionadded::1.9.2

**Comment by ali65** [comment]
I have also created some demo of the new feature: https://github.com/ali65/potplex
It would be nice to add this demo somehow to kivy examples. Any suggestions?

**Comment by ali65** [comment]
Demo here:
```python
from kivy.app import App
from kivy.uix.button import Button
from kivy import __version__ as kivy_version
from kivy.uix.boxlayout import BoxLayout
import os


#font_name = "/usr/share/fonts/truetype/msttcorefonts/arial.ttf"
# font_name = "/usr/share/fonts/truetype/msttcorefonts/Verdana.ttf"
# font_name = "/usr/share/fonts/truetype/roboto/Roboto-Regular.ttf"
font_name = "/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf"
fname = os.path.basename(font_name)
fsize = '13pt'


class TestApp(App):
    def build(self):
        layout = BoxLayout(orientation='vertical')
        layout.add_widget(
            Button(
                text="kivy_version: {}".format(kivy_version),
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='normal',
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text="Kerning On: AV",
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='normal',
                font_kerning=True,
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text="Kerning Off: AV",
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='normal',
                font_kerning=False,
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' solid none',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='none',
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' solid light',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='light',
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' solid normal',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='normal',
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' solid mono',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='mono',
                font_blended=False,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' blended none',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='none',
                font_blended=True,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' blended light',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='light',
                font_blended=True,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' blended normal',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='normal',
                font_blended=True,
                )
            )
        layout.add_widget(
            Button(
                text=fname + ' blended mono',
                background_normal='',
                color=[0, 0, 0, 1],
                font_name=font_name,
                font_size=fsize,
                font_hinting='mono',
                font_blended=True,
                )
            )
        return layout


def _main():
    TestApp().run()

if __name__ == '__main__':
    _main()
```
It would be nice to add this demo somehow to kivy examples. Any suggestions?

**Comment by dessant** [comment]
@ali65, a label example with the new options could be added here, in a separate pr: https://github.com/ali65/kivy/tree/master/examples/widgets.

**Comment by akshayaurora** [comment]
@ali65 thanks nice work !

**Comment by ali65** [comment]
Soon I will follow up with the merge request of the cleaned-up example of font rendering. Thanks for merging.

**Comment by dessant** [comment]
@akshayaurora :100: 



```python
from kivy.app import App
from kivy.lang import Builder

kv = """
Label:
    text: 'OFFICIALLY TESTED'
    font_size: '20dp'
"""


class TestApp(App):
    def build(self):
        return Builder.load_string(kv)

if __name__ == '__main__':
    TestApp().run()
```

**Comment by dessant** [comment]
This might also need to be sorted out:
```
matham: if TTF_STYLE_XXX is defined in the external headers, we should just do cdef int TTF_STYLE_XXX not cdef int TTF_STYLE_XXX = ...
```

**Pull request merged**